subsys/bluetooth/mesh/proxy_srv.c

```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+		req->duration = PROXY_RANDOM_UPDATE_INTERVAL;
+		return true;
+	}
+
+	if (bt_mesh_gatt_proxy_get() == BT_MESH_FEATURE_ENABLED) {
+		req->evt = NET_ID;
+		req->duration = SYS_FOREVER_MS;
+		return true;
+	}
+
+	return false;
+}
+
+static struct bt_mesh_subnet *adv_sub_get(struct bt_mesh_subnet *sub_start,
+					  struct proxy_adv_req *req, bool specific)
+{
+	struct bt_mesh_subnet *sub_temp = sub_start;
+
+	req->evt = NO_ADV;
+
+	do {
+		if (adv_sub_check(sub_temp, req)) {
+			return sub_temp;
+		}
+
+		sub_temp = bt_mesh_subnet_next(sub_temp);
+	} while (sub_temp != sub_start && !specific);
+
+	return NULL;
+}
+
+static int gatt_proxy_advertise(void)
 {
-	int32_t remaining = SYS_FOREVER_MS;
-	int subnet_count;
-	int err = -EBUSY;
-	bool planned = false;
+	int err;
+
+	int32_t max_adv_duration;
+	int active_sub_cnt;
+	static int32_t prev_start;
+
+	struct bt_mesh_subnet *sub;
+	struct proxy_adv_req request;
+	static struct bt_mesh_subnet *prev_sub;
+	static struct proxy_adv_req prev_req;
 
 	LOG_DBG("");
 
+	/* Close proxy activity that has timed out on all subnets */
+	bt_mesh_subnet_foreach(proxy_adv_check);
+
 	if (!bt_mesh_proxy_has_avail_conn()) {
 		LOG_DBG("Connectable advertising deferred (max connections)");
 		return -ENOMEM;
 	}
 
-	sub = beacon_sub ? beacon_sub : bt_mesh_subnet_next(beacon_sub);
-	if (!sub) {
-		LOG_WRN("No subnets to advertise on");
+	active_sub_cnt = active_proxy_sub_cnt();
+	if (!active_sub_cnt) {
+		LOG_DBG("No subnets to advertise proxy on");
 		return -ENOENT;
-	}
-
-	subnet_count = sub_count();
-	LOG_DBG("sub_count %u", subnet_count);
-	if (subnet_count > 1) {
-		int32_t max_timeout;
+	} else if (active_sub_cnt > 1) {
+		/** There are more than one subnet that requires proxy adv,
+		 *  and the adv resources must be shared.
+		 */
 
 		/* We use NODE_ID_TIMEOUT as a starting point since it may
 		 * be less than 60 seconds. Divide this period into at least
 		 * 6 slices, but make sure that a slice is at least one
 		 * second long (to avoid excessive rotation).
 		 */
-		max_timeout = NODE_ID_TIMEOUT / MAX(subnet_count, 6);
-		max_timeout = MAX(max_timeout, 1 * MSEC_PER_SEC);
+		max_adv_duration = NODE_ID_TIMEOUT / MAX(active_sub_cnt, 6);
+		max_adv_duration = MAX(max_adv_duration, MSEC_PER_SEC);
 
-		if (remaining > max_timeout || remaining == SYS_FOREVER_MS) {
-			remaining = max_timeout;
-		}
-	}
+		/* Check if the previous subnet finished its allocated timeslot */
+		if (is_sub_proxy_active(prev_sub) && prev_req.duration != SYS_FOREVER_MS) {
+			int32_t time_delta = k_uptime_get_32() - prev_start;
 
-	for (int i = 0; i < subnet_count; i++) {
+			if (time_delta < prev_req.duration) {
+				sub = adv_sub_get(prev_sub, &request, true);
 
-		if (sub->node_id == BT_MESH_NODE_IDENTITY_RUNNING) {
-			uint32_t active = k_uptime_get_32() - sub->node_id_start;
-			bool priv_node_id = false;
+				/* Check that it is the same event */
+				if ((sub == prev_sub) && (prev_req.evt == request.evt)) {
```

Comment:
Can't this call with `specific == true` be simply replaced with `adv_sub_check`? Because, unless I'm missing something, `while (sub_temp != sub_start && !specific)` exits immediately if `specific == true`. Then `adv_sub_get` either returns `prev_sub` or `NULL`.